### ISSUE TITLE
Consider allowing equality narrowing to ignore the possibility of a subclass overriding `__eq__`

### ISSUE THREAD

**Issue by clo-vis** [astral-sh/ty#3419]
### Summary

[Playground](https://play.ty.dev/8e172889-dc8d-4d77-b2d8-60641513c031)

```python
def g(p: str | None, q: str) -> None:
    s: str
    if p in [q]:
        s = p  # no error
    if p == q:
        s = p  # invalid-assignment
```

If `p in [q]` is True, ty infers that `p` cannot be None.
But if `p == q`, the type narrowing does not work.

**Comment by charliermarsh** [comment]
I think both of those narrowings are actually unsound.

**Comment by carljm** [comment]
> I think both of those narrowings are actually unsound.

They are, and we should handle them consistently: either decide to allow the unsound narrowing in both cases, or do it in neither.

What's unclear to me is whether it makes sense to allow the narrowing in #1566 but disallow this narrowing. It seems like this narrowing is only unsound when you actually have types defining `__eq__` so they compare equal to unrelated types, which is less common. #1566 is unsound even with a subclass that doesn't override `__eq__` at all. So it seems like we are more likely to hit unsoundness in practice in #1566. (But the unsoundness in #1566 may also be lower-stakes, in that it "just" results in wrongly typing a string subclass as the literal string it is equal to.)

Type checker behavior is split here, so there's no established consensus to follow. Pyright has the same oddly inconsistent behavior we currently do.

**Comment by AlexWaygood** [comment]
Our narrowing for `==` and `in` is now consistent following https://github.com/astral-sh/ruff/pull/26982, reflecting the fact that they use the same dunders at runtime. So the only remaining question here is whether we want to change them both so that we allow unsound union-equality narrowing (unless `tool.ty.strict-literal-narrowing` is set to `true`?).

**Comment by charliermarsh** [comment]
Is this the same as the questions we're considering elsewhere around (e.g.) assuming disjointness in certain cases via some `strict-subclass-narrowing` setting?

**Comment by charliermarsh** [comment]
(As in https://github.com/astral-sh/ruff/pull/26804 and https://github.com/astral-sh/ruff/pull/26415 and friends.)

**Comment by AlexWaygood** [comment]
yeah, it's a very similar idea to https://github.com/astral-sh/ruff/pull/26804, I think.

**Comment by carljm** [comment]
I think that we should change our equality narrowing to explicitly ignore the possibility of a subclass overriding `__eq__`. We've already in some ways gone further than this with #1566.

**Comment by charliermarsh** [comment]
That's fine with me
